### PR TITLE
Expose traceable logs

### PR DESCRIPTION
**Pull request by phillipj**
*Sorry about the big change set but logging cuts through most of the code*

ATM debugging what the bot does and errors isn't trivial for several reasons. We don't log enough about incoming events or Jenkins updates, and we don't have any way to trace whatever happened to one particular request.

It would also be nice to let more people have a look at the logs at any given time. Only having one person with access to them (which is me atm), isn't good enough.

These changes introduces bunyan for logging. Combined with an express middleware it'll add unique request ids to each incoming request, which we use throughout giving us the ability to search for everything that happened for any given request.

All logs are exposed and available for those who know the username / password combination set in `$LOGIN_CREDENTIALS`.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -5,16 +5,23 @@ require('dotenv').load({ silent: true })
...
+const bunyanMiddleware = require('bunyan-middleware')
+
+const logger = require('./lib/logger')
 
 const captureRaw = (req, res, buffer) => { req.raw = buffer }
 
 const app = express()
 app.use(bodyParser.json({ verify: captureRaw }))
+// bunyanMiddleware gives us request id's and unique loggers per incoming request,
+// for satefy reasons we don't want to include the webhook GitHub secret in logs
```
Comment (Fishrock123) [inline app.js]: Should we whitelist instead, or is that too impractical?

File: lib/github-events.js [hunk 2]
```diff
@@ -22,22 +22,28 @@ module.exports = (app) => {
 
     if (!signature || signature !== sign(secret, req.raw)) {
       res.writeHead(401, 'Invalid Signature')
-      console.error('! event@%s: %s - invalid signature, returning 401', data.repository.full_name, data.action)
+      req.log.error('Invalid GitHub event signature, returning 401')
```
Comment (Fishrock123) [inline lib/github-events.js]: where do you get `req.log`?

File: logs/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
Comment (Fishrock123) [inline logs/.gitignore]: newline please

**Comment by Fishrock123** [comment]
Seems fine, but tests aren't passing. I didn't 100% review all the variable passing.

**Comment by williamkapke** [comment]
What about using a service like Papertrail?

File: lib/github-events.js (hunk 2 above)
Comment (phillipj) [inline lib/github-events.js]: [bunyan-middleware](https://github.com/tellnes/bunyan-middleware) puts on every incoming request

File: app.js (hunk 1 above)
Comment (phillipj) [inline app.js]: ATM whitelisting is not implemented https://github.com/tellnes/bunyan-middleware#options. We could ofc open a PR if we really need it. Got any other sensitive headers we should avoid in the logs?

**Comment by phillipj** [comment]
> What about using a service like Papertrail?

I see this as a intermediate step between the `console.log`-ing we're doing today, and something like Papertrail. By using bunyan w/its cli we're able to get a lot of value of a very simple solution. Hopefully the bot will get deployed in the project's infra, idk what's preferred or not there in regards to such 3rd party services, which is also one of the reasons I'd like to keep it local and simpler at first.

File: lib/logger.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  streams: [
+    {
+      stream: process.stdout,
+      level: stdoutLevel
+    },
+    {
+      type: 'rotating-file',
+      path: path.join(__dirname, '../logs/bot.log'),
+      period: '1d', // daily rotation
+      count: 3      // keep three days of logging
```
Comment (jbergstroem) [inline lib/logger.js]: This looks like a very low default; especially if we'd be audit logging. Perhaps even avoid deletion? Optimally I'd like to see this passed as an option as well.

File: package.json [hunk 5]
```diff
@@ -3,7 +3,7 @@
   "version": "1.0.0-beta1",
   "description": "Node.js GitHub Bot",
   "scripts": {
-    "start": "node server.js",
+    "start": "node server.js | bunyan -o short",
```
Comment (jbergstroem) [inline package.json]: will this work as-is? `bunyan` wouldn't necessarily live in path; perhaps `node_modules/.bin/bunyan` ?

Comment (TheAlphaNerd) [inline package.json]: npm scripts shim in `node_modules/.bin` into the path :smile:

Comment (jbergstroem) [inline package.json]: oh. magic!

**Comment by phillipj** [comment]
Pushed an update which exposes the log files ^.

This completes my initial thoughts about exposing the logs to everyone who knows the user/pass, and maybe do some interesting queries on them:

```bash
# everything
$ curl http://example.org/logs/bot.log | bunyan -o short

# loglevel
$ curl http://example.org/logs/bot.log | bunyan -l error

# filter by PR #
$ curl http://example.org/logs/bot.log | bunyan -c 'this.pr == "nodejs/nodejs.org/#42"'

# filter by unique request
$ curl http://example.org/logs/bot.log | bunyan -c 'this.req_id == "some uuid here"'
```

File: lib/logger.js (hunk 4 above)
Comment (phillipj) [inline lib/logger.js]: Made env variable `$KEEP_LOGS` for configurability, defaults to 10 days.

**Comment by phillipj** [comment]
Merging this as I've addressed issues mentioned.

**Pull request merged**